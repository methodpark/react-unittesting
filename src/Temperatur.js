import React from 'react';

export default class Temperatur extends React.Component {
  render() {
    const {label, value} = this.props;

    return (
      <div id={label.toLowerCase()} className="temp">
        {label}: {this._round(value)}Â°C
      </div>
    );
  }

  _round(value) {
    return Math.round(value * 100) / 100;
  }
}