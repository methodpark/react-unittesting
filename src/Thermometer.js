import React, { Component } from 'react';

import Temperatur from './Temperatur';

export default class Thermometer extends Component {
  constructor() {
    super();
    this.state = {
      min: Infinity, 
      max: -Infinity
    };
  }

  static getDerivedStateFromProps(props, state) {
    return {
      min: Math.min(props.temp, state.min),
      max: Math.max(props.temp, state.max)
    };
  }

  render() {
    return (
      <div id="thermometer">
        <h1>Thermometer</h1>

        <Temperatur 
          label="Min" value={this.state.min} />
        <Temperatur
          label="Current" value={this.props.temp} />
        <Temperatur
          label="Max" value={this.state.max} />
      </div>
    );
  }
}