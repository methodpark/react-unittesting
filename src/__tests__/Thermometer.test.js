import React from 'react';
import { shallow, mount, render } from 'enzyme';

import Thermometer from '../Thermometer';
import Temperatur from '../Temperatur';

describe('render default thermo (mount)', () => {
  it('should render the min temperature', () => {
    const temperature = 20;

    const wrapper = mount(<Thermometer 
      temp={temperature} />);

    const minText = wrapper.find('#min').text();
    const expected = 'Min: 20°C';
    expect(minText).toEqual(expected);
  });

  it('should render the current temperature', () => {
    const temperature = 20;

    const wrapper = mount(<Thermometer 
      temp={temperature} />);

    const currentText = wrapper.find('#current').text();
    const expected = 'Current: 20°C';
    expect(currentText).toEqual(expected);
  });

  it('should render the max temperature', () => {
    const temperature = 20;

    const wrapper = mount(<Thermometer 
      temp={temperature} />);

    const maxText = wrapper.find('#max').text();
    const expected = 'Max: 20°C';
    expect(maxText).toEqual(expected);
  });
});

describe('render default thermo (shallow)', () => {
  it('should render the min temp', () => {
    const temperature = 20;
  
    const wrapper = shallow(<Thermometer
      temp={temperature} />);
  
    const min = wrapper.find(Temperatur).at(0);
    const expected = {label: 'Min', value: 20};
    expect(min.props()).toEqual(expected);
  });

  it('should render the current temp', () => {
    const temperature = 20;
  
    const wrapper = shallow(<Thermometer
      temp={temperature} />);
  
    const current = wrapper.find(Temperatur).at(1);
    const expected = {label: 'Current', value: 20};
    expect(current.props()).toEqual(expected);
  });

  it('should render the min temp', () => {
    const temperature = 20;
  
    const wrapper = shallow(<Thermometer
      temp={temperature} />);
  
    const max = wrapper.find(Temperatur).at(2);
    const expected = {label: 'Max', value: 20};
    expect(max.props()).toEqual(expected);
  });
});

describe('render thermo with updates (shallow)', () => {
  let wrapper = null;
  beforeAll(() => {
    wrapper = shallow(<Thermometer temp={20} />);
    wrapper.setProps({temp: 19.1211});
    wrapper.setProps({temp: 21.2221});
    wrapper.setProps({temp: 20.7123});
  });

  it('should render the min temperature', () => {
    const minTemp = wrapper.find(Temperatur).at(0);

    const expected = {label: 'Min', value: 19.1211};
    expect(minTemp.props()).toEqual(expected);
  });

  it('should render the current temperature', () => {
    const currentTemp = wrapper.find(Temperatur).at(1);

    const expected = {label: 'Current', value: 20.7123};
    expect(currentTemp.props()).toEqual(expected);
  });

  it('should render the max temperature', () => {
    const maxTemp = wrapper.find(Temperatur).at(2);

    const expected = {label: 'Max', value: 21.2221};
    expect(maxTemp.props()).toEqual(expected);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});