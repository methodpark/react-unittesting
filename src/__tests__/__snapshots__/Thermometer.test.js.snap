// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render thermo with updates (shallow) should match the snapshot 1`] = `
<div
  id="thermometer"
>
  <h1>
    Thermometer
  </h1>
  <Temperatur
    label="Min"
    value={19.1211}
  />
  <Temperatur
    label="Current"
    value={20.7123}
  />
  <Temperatur
    label="Max"
    value={21.2221}
  />
</div>
`;
